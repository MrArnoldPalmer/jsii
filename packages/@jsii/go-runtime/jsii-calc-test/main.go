package main

import (
	"fmt"
	baseofbase "github.com/aws-cdk/jsii/jsii-calc/golang/scopejsiicalcbaseofbase"
)

func main() {
	fmt.Println("Hello, JSII")

	// Class Creation
	number := baseofbase.NewVery()
        // Method invocation
	number.Hey()
}
