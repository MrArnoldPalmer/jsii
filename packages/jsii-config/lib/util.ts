import { readFile } from 'fs';

/* 
 * Look for existing nested values in config, return undefined if not found
 */
export function getNestedValue(keys: string[], current: object): any {
  try {
    return keys.reduce((val: any, key: string) => val[key], current);
  } catch (_err) {
    return undefined;
  }
}

/*
 * recursively flatten nested object
 */
export function flattenKeys(parent: string, fields: any) {
  return Object.entries(fields).reduce((accum: any, [key, vals]: [string, any]) => ({
    ...accum,
    [`${parent}.${key}`]: vals
  }), {});
}

/**
 * utility for for fs.readFile as a promise
 *
 * TODO: use util.promisify without breaking test stub
 */
export async function readFilePromise(path: string): Promise<Buffer> {
  const result: Buffer = await new Promise((resolve, reject) => {
    readFile(path, (err, data) => {
      if (err) {
        return reject(err);
      }
      return resolve(data);
    });
  });
  return result;
}
