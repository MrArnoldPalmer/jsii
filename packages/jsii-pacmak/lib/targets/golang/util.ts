// TODO: Generic utils for naming public/private data types & properties
export function goNameFromJs(name: string): string {
  return name.replace(/[^a-z0-9_.]/gi, '').toLowerCase();
}
