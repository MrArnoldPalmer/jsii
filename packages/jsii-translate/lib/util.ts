import fs = require('fs-extra');
import os = require('os');
import path = require('path');
import { renderTree, Source, translateTypeScript } from '.';
import { VisualizeAstVisitor } from './visitor';

export function startsWithUppercase(x: string) {
  return x.match(/^[A-Z]/);
}

export function inTempDir<T>(block: () => T): T {
  const origDir = process.cwd();
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'jsii'));
  process.chdir(tmpDir);
  const ret = block();
  process.chdir(origDir);
  fs.removeSync(tmpDir);
  return ret;
}

export function visualizeTypeScriptAst(source: Source) {
  const vis = translateTypeScript(source, new VisualizeAstVisitor(true), );
  return renderTree(vis.tree) + '\n';
}
