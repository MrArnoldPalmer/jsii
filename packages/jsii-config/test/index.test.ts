import * as fs from 'fs';
import * as inquirer from 'inquirer';
import buildQuestions from '../lib/questions';
import main from '../lib';
import { findQuestions, findQuestion } from './util';

describe('jsii-config', () => {
  describe('errors', () => {
    const readJsonMock = jest.fn();
    it('throws when package.json is invalid', async () => {
      readJsonMock.mockImplementation((_path, cb) => {
        cb(null, Buffer.from('INVALID JSON STRING'));
      });
      Object.defineProperty(fs, 'readFile', { value: readJsonMock });

      await expect(main()).rejects.toThrow('Unexpected token I in JSON at position 0');
    });
  });

  describe('no existing configuration', () => {
    const promptMock = jest.fn();
    const readJsonMock = jest.fn();
    const writeJsonMock = jest.fn();

    const packageJson = {
      key1: 'value1',
    };

    const config = {
      outdir: 'dist',
      versionFormat: 'short',
    };

    beforeEach(() => {
      promptMock.mockResolvedValueOnce({
        jsiiTargets: [],
        ...config
      }).mockResolvedValueOnce({
        confirm: true
      });

      readJsonMock.mockImplementation((_path, cb) => {
        cb(null, Buffer.from(JSON.stringify(packageJson)));
      });

      writeJsonMock.mockImplementation((_path, _data, cb) => {
        cb(null);
      });

      Object.defineProperty(fs, 'readFile', { value: readJsonMock });
      Object.defineProperty(inquirer, 'prompt', { value: promptMock });
      Object.defineProperty(fs, 'writeFile', { value: writeJsonMock });
    });

    afterEach(() => {
      promptMock.mockClear();
      readJsonMock.mockClear();
      writeJsonMock.mockClear();
    });

    afterAll(() => {
      promptMock.mockRestore();
      readJsonMock.mockRestore();
      writeJsonMock.mockClear();
    });

    it('prompts user for top level jsii config and language targets', async () => {
      await main();
      const questions = promptMock.mock.calls[0][0];
      const [outdir, versionFormat, targets] = findQuestions(['outdir', 'versionFormat', 'jsiiTargets'], questions);
      expect(outdir).toHaveProperty('type', 'input');
      expect(outdir).toHaveProperty('default', 'dist');

      expect(versionFormat).toHaveProperty('type', 'list');
      expect(versionFormat).toHaveProperty('choices', ['full', 'short']);

      expect(targets).toHaveProperty('type', 'checkbox');
      expect(targets).toHaveProperty('choices', ['java', 'python', 'dotnet']);
    });

    it('prompts for java specific values when only target enabled', async () => {
      await main();
      const enabled = { jsiiTargets: ['java'] };
      const disabled = { jsiiTargets: [] };
      const questions = promptMock.mock.calls[0][0];
      const subject = findQuestions([
        'targets.java.package',
        'targets.java.maven.groupId',
        'targets.java.maven.artifactId',
        'targets.java.maven.versionSuffix'
      ], questions);

      subject.forEach((question: any) => {
        expect(question.when(enabled)).toBe(true);
      });

      subject.forEach((question: any) => {
        expect(question.when(disabled)).toBe(false);
      });
    });

    it('prompts for python specific values only when target enabled', async () => {
      await main();
      const enabled = { jsiiTargets: ['python'] };
      const disabled = { jsiiTargets: [] };
      const questions = promptMock.mock.calls[0][0];
      const subject = findQuestions([
        'targets.python.module',
        'targets.python.distName'
      ], questions);

      subject.forEach((question: any) => {
        expect(question.when(enabled)).toBe(true);
      });

      subject.forEach((question: any) => {
        expect(question.when(disabled)).toBe(false);
      });
    });

    it('prompts for dotnet specific values only when target enabled', async () => {
      await main();
      const enabled = { jsiiTargets: ['dotnet'] };
      const disabled = { jsiiTargets: [] };
      const questions = promptMock.mock.calls[0][0];
      const subject = findQuestions([
        'targets.dotnet.namespace',
        'targets.dotnet.packageId',
        'targets.dotnet.iconUrl',
        'targets.dotnet.versionSuffix',
        'targets.dotnet.signAssembly'
      ], questions);

      subject.forEach((question: any) => {
        expect(question.when(enabled)).toBe(true);
      });

      subject.forEach((question: any) => {
        expect(question.when(disabled)).toBe(false);
      });
    });

    it('prompts for dotnet assembly originator file when target and signAssembly are enabled', async () => {
      await main();
      const questions = promptMock.mock.calls[0][0];
      const subject = findQuestion('targets.dotnet.assemblyOriginatorKeyFile', questions);
      const enabled = {
        jsiiTargets: ['dotnet'],
        targets: {
          dotnet: {
            signAssembly: true
          }
        }
      };

      const disabled = {
        jsiiTargets: ['dotnet'],
        targets: {
          dotnet: {
            signAssembly: false
          }
        }
      };

      expect(subject.when(enabled)).toBe(true);
      expect(subject.when(disabled)).toBe(false);
    });

    it('writes new config to package.json passed', async () => {
      await main();

      expect(writeJsonMock).toHaveBeenCalledWith('./package.json', JSON.stringify({
        ...packageJson,
        jsii: config
      }, null, 2), expect.anything());
    });
  });

  describe('existing configuration', () => {
    const promptMock = jest.fn();
    const readJsonMock = jest.fn();
    const writeJsonMock = jest.fn();
    
    const existingConfig = {
      outdir: 'OUTDIR',
      versionFormat: 'short',
      targets: {
        java: {
          package: 'JAVA_PACKAGE',
          maven: {
            groupId: 'JAVA_MAVEN_GROUPID',
            artifactId: 'JAVA_MAVEN_ARTIFACTID'
          }
        },
        dotnet: {
          namespace: 'DOTNET_NAMESPACE',
          packageId: 'DOTNET_PACKAGEID',
          iconUrl: 'DOTNET_ICONURL'
        },
        python: {
          distName: 'PYTHON_DISTNAME',
          module: 'PYTHON_MODULE'
        }
      },
      metadata: {
        'jsii:boolean': true,
        'jsii:number': 1337
      }
    };

    const config = {
      outdir: 'dist',
      versionFormat: 'short'
    };

    beforeEach(() => {
      promptMock
        .mockResolvedValueOnce(config)
        .mockResolvedValueOnce({
          confirm: true
        });

      readJsonMock.mockImplementation((_path, cb) => {
        cb(null, Buffer.from(JSON.stringify({
          jsii: existingConfig
        })));
      });

      writeJsonMock.mockImplementation((_path, _data, cb) => {
        cb(null);
      });

      Object.defineProperty(fs, 'readFile', { value: readJsonMock });
      Object.defineProperty(inquirer, 'prompt', { value: promptMock });
      Object.defineProperty(fs, 'writeFile', { value: writeJsonMock });
    });

    afterEach(() => {
      promptMock.mockClear();
      readJsonMock.mockClear();
      writeJsonMock.mockClear();
    });

    afterAll(() => {
      promptMock.mockRestore();
      readJsonMock.mockRestore();
      writeJsonMock.mockClear();
    });

    it('uses existing values as prompt defaults', async () => {
      await main();
      const defaultMap: { [key: string]: any } = {
        outdir: 'OUTDIR',
        versionFormat: 'short',
        jsiiTargets: ['java', 'dotnet', 'python'],
        ['targets.java.package']: 'JAVA_PACKAGE',
        ['targets.java.maven.groupId']: 'JAVA_MAVEN_GROUPID',
        ['targets.java.maven.artifactId']: 'JAVA_MAVEN_ARTIFACTID',
        ['targets.python.distName']: 'PYTHON_DISTNAME',
        ['targets.python.module']: 'PYTHON_MODULE',
        ['targets.dotnet.namespace']: 'DOTNET_NAMESPACE',
        ['targets.dotnet.packageId']: 'DOTNET_PACKAGEID',
        ['targets.dotnet.iconUrl']: 'DOTNET_ICONURL'
      };
      const questions = promptMock.mock.calls[0][0];

      Object.entries(defaultMap).forEach((entry: [string, any]) => {
        const [name, defaultVal] = entry;
        const subject = findQuestion(name, questions);

        expect(subject).toHaveProperty('default', defaultVal);
      });
    });

    it('preserves existing jsii metadata fields', async () => {
      await main();
      expect(writeJsonMock).toHaveBeenCalledWith('./package.json', JSON.stringify({
        jsii: {
          ...config,
          metadata: {
            'jsii:boolean': true,
            'jsii:number': 1337
          }
        }
      }, null, 2), expect.anything());
    });
  });

  describe('edit config', () => {
    const promptMock = jest.fn();
    const readJsonMock = jest.fn();
    const writeJsonMock = jest.fn();
    const answers = {
      jsiiTargets: ['python'],
      outdir: 'OUTDIR',
      versionFormat: 'short',
      targets: {
        python: {
          distName: 'PYTHON_DISTNAME',
          module: 'PYTHON_MODULE'
        }
      }
    };

    beforeEach(() => {
      promptMock
        .mockResolvedValueOnce(answers)
        .mockResolvedValueOnce({
          confirm: false
        })
        .mockResolvedValueOnce(answers)
        .mockResolvedValueOnce({
          confirm: true
        });

      readJsonMock.mockImplementation((_path, cb) => {
        cb(null, Buffer.from(JSON.stringify({})));
      });

      writeJsonMock.mockImplementation((_path, _data, cb) => {
        cb(null);
      });

      Object.defineProperty(fs, 'readFile', { value: readJsonMock });
      Object.defineProperty(inquirer, 'prompt', { value: promptMock });
      Object.defineProperty(fs, 'writeFile', { value: writeJsonMock });
    });

    afterEach(() => {
      promptMock.mockClear();
      readJsonMock.mockClear();
      writeJsonMock.mockClear();
    });

    afterAll(() => {
      promptMock.mockRestore();
      readJsonMock.mockRestore();
      writeJsonMock.mockRestore();
    });

    it('prompts user again with previous answers if confirmation is declined', async () => {
      await main();
      const defaultMap: { [key: string]: any } = {
        outdir: 'OUTDIR',
        versionFormat: 'short',
        jsiiTargets: ['python'],
        ['targets.python.distName']: 'PYTHON_DISTNAME',
        ['targets.python.module']: 'PYTHON_MODULE',
      };
      const questions = promptMock.mock.calls[2][0];

      Object.entries(defaultMap).forEach((entry: [string, any]) => {
        const [name, defaultVal] = entry;
        const subject = findQuestion(name, questions);

        expect(subject).toHaveProperty('default', defaultVal);
      });
    });
  });

  describe('snapshots', () => {
    it('builds prompt questions with no current config', () => {
      const subject = buildQuestions();
      expect(subject).toMatchSnapshot();
    });

    it('builds prompt questions with current config values as defaults', () => {
      const subject = buildQuestions({
        outdir: 'dist',
        targets: {
          java: {
            package: 'software.amazon.jsii.tests.calculator.base',
            maven: {
              groupId: 'software.amazon.jsii.tests',
              artifactId: 'calculator-base'
            }
          },
          dotnet: {
            namespace: 'Amazon.JSII.Tests.CalculatorNamespace.BaseNamespace',
            packageId: 'Amazon.JSII.Tests.CalculatorPackageId.BasePackageId'
          },
          python: {
            distName: 'scope.jsii-calc-base',
            module: 'scope.jsii_calc_base'
          }
        },
        versionFormat: 'short'
      });
      expect(subject).toMatchSnapshot();
    });
  });
});
