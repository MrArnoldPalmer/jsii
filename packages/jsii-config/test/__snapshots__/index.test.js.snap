// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsii-config snapshots builds prompt questions with current config values as defaults 1`] = `
Array [
  Object {
    "choices": Array [
      "java",
      "python",
      "dotnet",
    ],
    "default": Array [
      "java",
      "dotnet",
      "python",
    ],
    "message": "Target Languages",
    "name": "jsiiTargets",
    "type": "checkbox",
  },
  Object {
    "choices": Array [
      "deprecated",
      "experimental",
      "stable",
      "external",
    ],
    "default": "experimental",
    "message": "Jsii Stability - stability of compiled module apis",
    "name": "stability",
    "type": "list",
  },
  Object {
    "message": "Jsii Type Definitions - compiled typescript definitions file for module (e.g. index.d.ts)",
    "name": "types",
    "type": "input",
    "validate": [Function],
  },
  Object {
    "default": "dist",
    "message": "Output directory for typescript compiler (e.g. dist)",
    "name": "jsii.outdir",
    "type": "input",
    "validate": [Function],
  },
  Object {
    "choices": Array [
      "full",
      "short",
    ],
    "default": "short",
    "message": "Version format, default is full",
    "name": "jsii.versionFormat",
    "type": "list",
  },
  Object {
    "default": "software.amazon.jsii.tests.calculator.base",
    "message": "Java Package - root java package name under which the types will be declared (e.g. software.amazon.module.core)",
    "name": "jsii.targets.java.package",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "default": "software.amazon.jsii.tests",
    "message": "Maven GroupID - package group id (e.g. software.amazon.module)",
    "name": "jsii.targets.java.maven.groupId",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "default": "calculator-base",
    "message": "Maven ArtifactID - package artifact id (e.g. core)",
    "name": "jsii.targets.java.maven.artifactId",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "filter": [Function],
    "message": "Maven Version Suffix - optional suffix appended to the end of the maven package's version field (e.g. .DEVPREVIEW)",
    "name": "jsii.targets.java.maven.versionSuffix",
    "type": "input",
    "when": [Function],
  },
  Object {
    "default": "scope.jsii-calc-base",
    "message": "Python Distname - PyPI distribution name for the package (e.g. module-name.core)",
    "name": "jsii.targets.python.distName",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "default": "scope.jsii_calc_base",
    "message": "Python Module - name of the generated Python module (e.g. module_name.core)",
    "name": "jsii.targets.python.module",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "default": "Amazon.JSII.Tests.CalculatorNamespace.BaseNamespace",
    "message": ".NET Namespace - root namespace under which types will be declared (e.g. Amazon.Module)",
    "name": "jsii.targets.dotnet.namespace",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "default": "Amazon.JSII.Tests.CalculatorPackageId.BasePackageId",
    "message": ".NET Package Id - identifier of the package in the NuGet registry (e.g. Amazon.Module)",
    "name": "jsii.targets.dotnet.packageId",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "filter": [Function],
    "message": ".NET Icon Url - optional url of the icon to be shown in the NuGet gallery (e.g. https://raw.githubusercontent.com/module-icon.png)",
    "name": "jsii.targets.dotnet.iconUrl",
    "type": "input",
    "when": [Function],
  },
  Object {
    "default": "",
    "filter": [Function],
    "message": ".NET Version Suffix - optional suffix that will be appended at the end of the NuGet package's version field, must begin with a \\"-\\" (e.g. -devpreview)",
    "name": "jsii.targets.dotnet.versionSuffix",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "default": false,
    "message": ".NET Sign Assembly - whether the assembly should be strong-name signed. Defaults to false when not specified",
    "name": "jsii.targets.dotnet.signAssembly",
    "type": "confirm",
    "when": [Function],
  },
  Object {
    "default": "",
    "message": ".NET Assembly Originator Key File - path to the strong-name signing key to be used (e.g. ../../key.snk)",
    "name": "jsii.targets.dotnet.assemblyOriginatorKeyFile",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
]
`;

exports[`jsii-config snapshots builds prompt questions with no current jsii config 1`] = `
Array [
  Object {
    "choices": Array [
      "java",
      "python",
      "dotnet",
    ],
    "default": Array [],
    "message": "Target Languages",
    "name": "jsiiTargets",
    "type": "checkbox",
  },
  Object {
    "choices": Array [
      "deprecated",
      "experimental",
      "stable",
      "external",
    ],
    "default": "experimental",
    "message": "Jsii Stability - stability of compiled module apis",
    "name": "stability",
    "type": "list",
  },
  Object {
    "message": "Jsii Type Definitions - compiled typescript definitions file for module (e.g. index.d.ts)",
    "name": "types",
    "type": "input",
    "validate": [Function],
  },
  Object {
    "default": "dist",
    "message": "Output directory for typescript compiler (e.g. dist)",
    "name": "jsii.outdir",
    "type": "input",
    "validate": [Function],
  },
  Object {
    "choices": Array [
      "full",
      "short",
    ],
    "default": "full",
    "message": "Version format, default is full",
    "name": "jsii.versionFormat",
    "type": "list",
  },
  Object {
    "message": "Java Package - root java package name under which the types will be declared (e.g. software.amazon.module.core)",
    "name": "jsii.targets.java.package",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "message": "Maven GroupID - package group id (e.g. software.amazon.module)",
    "name": "jsii.targets.java.maven.groupId",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "message": "Maven ArtifactID - package artifact id (e.g. core)",
    "name": "jsii.targets.java.maven.artifactId",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "filter": [Function],
    "message": "Maven Version Suffix - optional suffix appended to the end of the maven package's version field (e.g. .DEVPREVIEW)",
    "name": "jsii.targets.java.maven.versionSuffix",
    "type": "input",
    "when": [Function],
  },
  Object {
    "message": "Python Distname - PyPI distribution name for the package (e.g. module-name.core)",
    "name": "jsii.targets.python.distName",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "message": "Python Module - name of the generated Python module (e.g. module_name.core)",
    "name": "jsii.targets.python.module",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "message": ".NET Namespace - root namespace under which types will be declared (e.g. Amazon.Module)",
    "name": "jsii.targets.dotnet.namespace",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "message": ".NET Package Id - identifier of the package in the NuGet registry (e.g. Amazon.Module)",
    "name": "jsii.targets.dotnet.packageId",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "filter": [Function],
    "message": ".NET Icon Url - optional url of the icon to be shown in the NuGet gallery (e.g. https://raw.githubusercontent.com/module-icon.png)",
    "name": "jsii.targets.dotnet.iconUrl",
    "type": "input",
    "when": [Function],
  },
  Object {
    "default": "",
    "filter": [Function],
    "message": ".NET Version Suffix - optional suffix that will be appended at the end of the NuGet package's version field, must begin with a \\"-\\" (e.g. -devpreview)",
    "name": "jsii.targets.dotnet.versionSuffix",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
  Object {
    "default": false,
    "message": ".NET Sign Assembly - whether the assembly should be strong-name signed. Defaults to false when not specified",
    "name": "jsii.targets.dotnet.signAssembly",
    "type": "confirm",
    "when": [Function],
  },
  Object {
    "default": "",
    "message": ".NET Assembly Originator Key File - path to the strong-name signing key to be used (e.g. ../../key.snk)",
    "name": "jsii.targets.dotnet.assemblyOriginatorKeyFile",
    "type": "input",
    "validate": [Function],
    "when": [Function],
  },
]
`;
